### PR TITLE
[Feature] Supporting Query Filters. Associations refactoring.

### LINKED ISSUE #860: Member is not table column
I use OData WebApi Library over LinqToDb. 
Library generate expression for select all records where association contains entities satisfy some condition.
LinqToDb rise exception.
```
LinqToDB.Linq.LinqException occurred
  HResult=0x80131500
  Message=Member 'b.C' is not a table column.
  StackTrace:
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.GetField(Expression expression, Int32 level, Boolean throwException) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\TableBuilder.cs:line 1204
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.FindTable(Expression expression, Int32 level, Boolean throwException, Boolean throwExceptionForNull) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\TableBuilder.cs:line 1237
   at LinqToDB.Linq.Builder.TableBuilder.TableContext.ConvertToSql(Expression expression, Int32 level, ConvertFlags flags) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\TableBuilder.cs:line 706
   at LinqToDB.Linq.Builder.ExpressionContext.ConvertToSql(Expression expression, Int32 level, ConvertFlags flags) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionContext.cs:line 57
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 840
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 797
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertCompare(IBuildContext context, ExpressionType nodeType, Expression left, Expression right) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1583
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertPredicate(IBuildContext context, Expression expression) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1363
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSearchCondition(IBuildContext context, Expression expression, List`1 conditions) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 2503
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 672
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 795
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertCompare(IBuildContext context, ExpressionType nodeType, Expression left, Expression right) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1583
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertPredicate(IBuildContext context, Expression expression) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1363
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSearchCondition(IBuildContext context, Expression expression, List`1 conditions) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 2503
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildWhere(IBuildContext parent, IBuildContext sequence, LambdaExpression condition, Boolean checkForSubQuery, Boolean enforceHaving) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 45
   at LinqToDB.Linq.Builder.AllAnyBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\AllAnyBuilder.cs:line 33
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\MethodCallBuilder.cs:line 22
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.cs:line 183
   at LinqToDB.Linq.Builder.ExpressionBuilder.GetSubQuery(IBuildContext context, MethodCallExpression expr) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 238
   at LinqToDB.Linq.Builder.ExpressionBuilder.SubQueryToSql(IBuildContext context, MethodCallExpression expression) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 252
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 879
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 763
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertToSql(IBuildContext context, Expression expression, Boolean unwrap) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 797
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertCompare(IBuildContext context, ExpressionType nodeType, Expression left, Expression right) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1583
   at LinqToDB.Linq.Builder.ExpressionBuilder.ConvertPredicate(IBuildContext context, Expression expression) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 1363
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSearchCondition(IBuildContext context, Expression expression, List`1 conditions) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 2503
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildWhere(IBuildContext parent, IBuildContext sequence, LambdaExpression condition, Boolean checkForSubQuery, Boolean enforceHaving) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.SqlBuilder.cs:line 45
   at LinqToDB.Linq.Builder.WhereBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\WhereBuilder.cs:line 26
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\MethodCallBuilder.cs:line 22
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.cs:line 183
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]() in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Builder\ExpressionBuilder.cs:line 154
   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Query.cs:line 261
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\Query.cs:line 210
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache) in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\ExpressionQuery.cs:line 84
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator() in C:\Users\genus\Projects\Repos\linq2db\Source\Linq\ExpressionQuery.cs:line 186
   at System.Collections.Generic.LargeArrayBuilder`1.AddRange(IEnumerable`1 items)
   at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at ConsoleApp1.Program.Main(String[] args) in C:\Users\genus\Projects\Repos\linq2db\ConsoleApp1\Program.cs:line 33

```

### Steps to reproduce
```c#
class Program
{
        static void Main(string[] args)
        {
		using (var cx = new DataContext(new LinqToDB.DataProvider.SqlServer.SqlServerDataProvider("SqlServer", LinqToDB.DataProvider.SqlServer.SqlServerVersion.v2008), "Server=(localdb)\\MSSQLLocalDB;Database=Satellite;Integrated security=true"))
		{
			var q = cx.GetTable<a>()
				.Where(it => (
						(it.B == null)
							?null
							:((bool?)it.B.Any(d => 
								(
									((d == null ? null : d.C) == null)
										? null
										:d.C.prop
								) == "aaa")
							)
					) == true
				);
			q.ToArray();
		}
        }
    }
	[Table]
	class a {
		[Column]
		public int Id { get; set; }
		[Column]
		public int bId { get; set; }
		[Association(ThisKey =nameof(bId), OtherKey =nameof(b.Id))]
		public IList<b> B { get; set; }
	}

	[Table]
	class b
	{
		[Column]
		public int Id { get; set; }
		[Column]
		public int cId { get; set; }
		[Association(ThisKey = nameof(cId), OtherKey = nameof(c.Id))]
		public c C { get; set; }
	}

	[Table]
	class c
	{
		[Column]
		public int Id { get; set; }
		[Column]
		public string prop { get; set; }
	}
```

### Environment details
linq2db version: 1.9.0
Database Server: Sql Server LocalDB
Operating system: Windows 10  
Framework version: .NET Core 2.0

### LINKED ISSUE #1711: Association using a query expression throws a cast exception when using an IQueryable property.
I'm attempting to add additional query constraints to an association via a query expression. Additionally, when using this association in a query, I would like to perform a .Any() linq query upon this association.

The following configuration works:
```
Mapping:
builder
	.Entity<Entity>()
	.HasTableName("Entity")
	.HasPrimaryKey(x => Sql.Property<long>(x, "Id"))
	.Association(x => Sql.Property<IQueryable<Relationship>>(x, "relationship"),
		e => e.Id, r => r.EntityId);
```

```
Query:
var result1 = entity
	.Where(t => Sql.Property<IQueryable<Relationship>>(t, "relationship").Any())
	.ToList();

SQL:
SELECT
	[t].[Id],
	[t].[Deleted],
	....
FROM
	[Entity] [t]
WHERE
	EXISTS(
		SELECT
			*
		FROM
			[Relationship] [c_1]
		WHERE
			[t].[Id] = [c_1].[EntityId]
	)
```

The following configuration fails with an exception when using the same query from above:
```
builder
	.Entity<Entity>()
	.HasTableName("Entity")
	.HasPrimaryKey(x => Sql.Property<long>(x, "Id"))
	.Association(x => Sql.Property<IQueryable<Relationship>>(x, "relationship"),
		(e, db) => db.GetTable<Relationship>()
		    .Where(r => r.Deleted == false && r.EntityId == e.Id));
```

If you are seeing an exception, include the full exceptions details (message and stack trace).

```
System.ArgumentException: 'Expression of type 'Linq2DBTest.Relationship' cannot be used for parameter of type 'System.Linq.IQueryable`1[Linq2DBTest.Relationship]' of method 'Boolean Any[Relationship](System.Linq.IQueryable`1[Linq2DBTest.Relationship])'
Parameter name: arg0'
```

### Environment details
linq2db version: *2.7.0*
Database Server: *E.g. SQL Server 2017*
Database Provider: *E.g. SqlClient*
Operating system: *E.g. Windows 10*
Framework version: *.NET Core 2.2*

### LINKED ISSUE #1906: INNER JOIN nested inside a LEFT JOIN => Join Expression not supported exception
### Problem
I try return from static method a complete object
```
private static IQueryable<CtqResultModel> GetFull(this DbQms db)
{
    return db.CtqResults.
        LoadWith(f => f.Definition.Set.Sector).
        AsQueryable();
}
```
Result belongs to 1 Definition. Definition belongs to 0 or 1 Set. Set belongs to 1 Sector.

Produced query:
```
SELECT
*
FROM
(([tblCTQResult] [r]
    INNER JOIN [tblCTQDefinition] [db_Definition] ON ([r].[DefinitionId] = [db_Definition].[DefinitionId]))
    LEFT JOIN [tblCTQSet] [db_Set] ON ([db_Definition].[SetId] = [db_Set].[SetId]))
	INNER JOIN [tblSector] [db_Sector] ON ([db_Set].[SectorId] = [db_Sector].[Id])
```
gives
`System.Data.OleDb.OleDbException: 'JOIN expression not supported.'`

I found two sources where this problem is mentioned:
1. Due to [limitation in MS Access](https://docs.microsoft.com/en-us/previous-versions/office/developer/office-2007/bb208854(v=office.12)?redirectedfrom=MSDN) a LEFT JOIN or a RIGHT JOIN may be nested inside an INNER JOIN, but an INNER JOIN may not be nested inside a LEFT JOIN or a RIGHT JOIN.
2. Here I found [proposition](https://stackoverflow.com/a/28685893) to convert the second join to a left join.

How to solve this issue? Do you know any workaround?

### Steps to reproduce
```
public class CtqResultModel
{
    [Column, PrimaryKey, Identity]
    public int ResultId { get; set; }
    
    [Column, NotNull]
    public int DefinitionId { get; set; }
    
    [Association(ThisKey = nameof(DefinitionId), OtherKey = nameof(CtqDefinitionModel.DefinitionId), CanBeNull = false)]
    public CtqDefinitionModel Definition { get; set; }
}

public class CtqDefinitionModel
{
    [Column, PrimaryKey, Identity]
    public int DefinitionId { get; set; }
    
    [Column, NotNull]
    public int SetId { get; set; }
    
    [Association(ThisKey = nameof(SetId), OtherKey = nameof(CtqSetModel.SetId), CanBeNull = true)]
    public CtqSetModel Set { get; set; }
}

public class CtqSetModel
{
    [Column, PrimaryKey, Identity]
    public int SetId { get; set; }
    
    [Column, NotNull]
    public int SectorId { get; set; }
    
    [Association(ThisKey = nameof(SectorId), OtherKey = nameof(FtqSectorModel.Id), CanBeNull = false)]
    public FtqSectorModel Sector { get; set; }
}

public class FtqSectorModel
{
    [Column, PrimaryKey, Identity]
    public int Id { get; set; }
}
```
### Environment details
linq2db version: 2.9.0
Database Server: MS Access
Database Provider: OleDb
Operating system: Windows10
.NET Framework: 4.5

### LINKED ISSUE #2199: Bug with fix
Take these two classes:
```

[Table("Manufacturer")]
public class Manufacturer
{
    [Column]
    [PrimaryKey]
    public int ManufacturerId { get; set; }

    [Column]
    public string Name { get; set; }

    [Column]
    public string CountryCode { get; set; }

    [Association(ThisKey = nameof(CountryCode), OtherKey = nameof(Country.Code))]
    public Country CountryEntity { get; set; }
}

[Table("Country")]
public class Country
{
    [Column]
    [PrimaryKey]
    public string Code { get; set; }

    [Column]
    public string Name { get; set; }
}
```
```

var man1 = from e in _context.GetTable<Manufacturer>()
                               join c in _context.GetTable<Country>() on e.CountryCode equals c.Code into ce
                               from co in ce.DefaultIfEmpty()
                               where e.ManufacturerId == 1
                               select new Manufacturer { ManufacturerId=e.ManufacturerId, CountryCode=e.CountryCode,Name=e.Name,CountryEntity=co};

            var man2 = from e in _context.GetTable<Manufacturer>()
                       join c in _context.GetTable<Country>() on e.CountryCode equals c.Code into ce
                       from co in ce.DefaultIfEmpty()
                       where e.ManufacturerId == 2
                       select new Manufacturer { ManufacturerId = e.ManufacturerId, CountryCode = e.CountryCode, Name = e.Name, CountryEntity = co };

            var query1 = from m1 in man1
                               join m2 in man2 on 1 equals 1 into manjoin
                               from m in manjoin.DefaultIfEmpty()
                               select new Tuple<Manufacturer, Manufacturer>(m1, m);

            var query2 = from m1 in man1
                         from m2 in man2
                         select new Tuple<Manufacturer, Manufacturer>(m1, m2);
```
This is much simplified version of one of our queries.
Query1 is not correctly translated:
```

SELECT
	[e_2].[ManufacturerId],
	[e_2].[CountryCode],
	[e_2].[Name],
	[t1].[co],
	[t1].[Name],
	[t1].[c1],
	[t3].[ManufacturerId],
	[t3].[CountryCode],
	[t3].[Name],
	[ce_2].[Code],
	[ce_2].[Name],
	[t2].[c1],
	[t3].[c1]
FROM
	[Manufacturer] [e_2]
		LEFT JOIN (
			SELECT
				[ce_1].[Code] as [co],
				1 as [c1],
				[ce_1].[Name]
			FROM
				[Country] [ce_1]
		) [t1] ON ([e_2].[CountryCode] IS NULL AND [t1].[co] IS NULL OR [e_2].[CountryCode] = [t1].[co])
		LEFT JOIN (
			SELECT
				1 as [c1],
				[e_1].[ManufacturerId],
				[e_1].[CountryCode],
				[e_1].[Name]
			FROM
				[Manufacturer] [e_1]
					LEFT JOIN (
						SELECT
							[ce_2].[Code] as [co],
							1 as [c1]
						FROM
							[Country] [ce_2]
					) [t2] ON ([e_1].[CountryCode] IS NULL AND [t2].[co] IS NULL OR [e_1].[CountryCode] = [t2].[co])
			WHERE
				[e_1].[ManufacturerId] = 2
		) [t3] ON 1=1
WHERE
	[e_2].[ManufacturerId] = 1

```
Error is:

```
The multi-part identifier "ce_2.Code" could not be bound.
The multi-part identifier "ce_2.Name" could not be bound.
The multi-part identifier "t2.c1" could not be bound.
```
You are taking alias from subquery.

I replaced query1 with query2 and that one works.
This isn't critical bug for us, but it is worth investigating. This is valid query and it worked on EF.

### Environment details
linq2db version: 3.0.0.preview2
Database Server: SqlServer 2016
Database Provider: MicrosoftDataSqlClient
Operating system: Windows 10
.NET Framework: .Net Core 3.1.3

### PR DESCRIPTION
**Pull request by sdanyliv**
Added supporting of query filters and refactored associations support.
Added support of GROUP BY: GROUPING SETS, ROLLUP, CUBE.

Fixes #2199 
Fixes #1906
Fixes #1711 
Fixes #860

Tasks to TODO

- [x] Passing Merge tests
- [x] More filter tests
- [x] More LoadWith overloads
- [x] Cache hits with Query Filters
- [x] Code cleanup
- [x] Update XML documentation
- [x] Cleanup Playground project

### CODE REVIEW

File: Source/LinqToDB/Linq/Builder/ContextRefBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+		public bool CanBuild(ExpressionBuilder builder, BuildInfo buildInfo)
+		{
+			return buildInfo.Expression is ContextRefExpression;
+		}
+
+		public IBuildContext BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
+		{
+			var context = ((ContextRefExpression)buildInfo.Expression).BuildContext;
+			if (buildInfo.IsSubQuery && buildInfo.SelectQuery.From.Tables.Count == 0)
```
Comment (MaceWindu) [inline Source/LinqToDB/Linq/Builder/ContextRefBuilder.cs:18]: ?

File: Source/LinqToDB/Expressions/TransformInfo.cs [hunk 2]
```diff
@@ -8,15 +8,26 @@ public TransformInfo(Expression expression, bool stop)
...
+		{
+			Expression = expression;
+			Stop       = false;
+			Stop       = stop;
+			Continue   = @continue;
 		}
 
 		public Expression Expression;
 		public bool       Stop;
+		public bool		  Continue;
```
Comment (MaceWindu) [inline Source/LinqToDB/Expressions/TransformInfo.cs:31]: formatting

File: Source/LinqToDB/LinqToDB.csproj [hunk 3]
```diff
@@ -14,7 +14,7 @@
 	</ItemGroup>
 
 	<PropertyGroup Condition="'$(Configuration)'=='Debug'">
-		<DefineConstants>OVERRIDETOSTRING1;$(DefineConstants)</DefineConstants>
+		<DefineConstants>OVERRIDETOSTRING;$(DefineConstants)</DefineConstants>
```
[the remark this answers is not in the document]
Comment (sdanyliv) [reply Source/LinqToDB/LinqToDB.csproj:17]: I think we can keep this option on. Looks like i have fixed stack overflow problems here.

File: Source/LinqToDB/Sql/Sql.GroupBy.cs [hunk 4]
```diff
@@ -1,13 +1,45 @@
...
+			{
+				throw new LinqException($"'{nameof(GroupingSets)}' should not be called directly.");
+			}
 		}
+
+		public static IGroupBy GroupBy = new GroupByImpl();
+
+		[Sql.Extension("GROUPING({field})", ServerSideOnly = true, CanBeNull = false, IsAggregate = true)]
+		public static int Grouping<T>([ExprParameter] T field) 
+			=> throw new LinqException($"'{nameof(GroupBy)}' should not be called directly.");
```
[the remark this answers is not in the document]
Comment (sdanyliv) [reply Source/LinqToDB/Sql/Sql.GroupBy.cs:42]: Grouping is a special function which checks that specific property currently in grouping process.
What i have missed here is XML documentation.

**Pull request merged**